### PR TITLE
bash-completion: Adapt for 0.12 and 0.13

### PR DESCRIPTION
**Pull request by roques**
- separate completion for bitcoind and bitcoin-cli
- remove RPC support from bitcoind completion
- add completion for bitcoin-tx and bitcoin-qt

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Nice! Thanks.
Concept ACK.

File: contrib/bitcoin-tx.bash-completion [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# bash programmable completion for bitcoin-tx(1)
+1<# Copyright (c) 2016 Christian von Roques <roques@mti.ag>
```
Comment (MarcoFalke) [inline contrib/bitcoin-tx.bash-completion]: Are the first two chars desired?

Comment (roques) [inline contrib/bitcoin-tx.bash-completion]: Definitely not. Somehow I too often hit «1» instead of «esc» these days.

**Comment by luke-jr** [comment]
Concept ACK. Would you mind switching to the standard project copyright header while you're at it?

**Comment by luke-jr** [comment]
The "Copyright Bitcoin Core developers" is essentially saying each author retains copyright to their own contributions; there is no centralised/formal/legal entity. Maybe we should clarify this somewhere. (To be clear, I don't consider this a blocker to your PR, just a "nice to have".)

**Comment by MarcoFalke** [comment]
If you are the sole author of the file, it is fine to just put your name and not "Bitcoin Core developers". However, if someone modifies the file, we just put "Bitcoin Core developers" in there to not clutter the headers. The detailed list of authors could then be fetched via `git blame` or `git log` or https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.10.0.md#credits

**Comment by roques** [comment]
I did some more digging through the Copyright headers. Below `src/` the situation is very clear: Either a file originates with Bitcoin and is "Copyright (c) The Bitcoin Core developers" or it originates from an external project and keeps that project's authors' Copyright headers. Satoshi is an exception to that rule. Outside of `src/`, especially in `contrib/devtools/` there are several files which retain their initial authors' Copyright.
I like @luke-jr's interpretation and second the wish for an explicit mention of it somewhere in the developer documentation. — My initial resistance was due to me misunderstanding that I was being asked to assign my copyright to somebody else.

**Comment by MarcoFalke** [comment]
Travis failure unrelated in tests.exe:

```
Application tried to create a window, but no driver could be loaded.
Make sure that your X server is running and that $DISPLAY is set correctly.
err:systray:initialize_systray Could not create tray window
test count = err:seh:setup_exception stack overflow 3648 bytes in thread 0026 eip 00002aef24a23efb esp 00000000001407c0 stack 0x140000-0x141000-0x340000

**Comment by petertodd** [comment]
The phrase "Bitcoin Core developers" in the copyright notice just means "developers of Bitcoin Core" - which is anyone at all who has contributed anything. There's a good writeup on this subject that here, which takes into account how distributed revision control systems like git do a good job of recording exact authorship for legal purposes: https://softwarefreedom.org/resources/2012/ManagingCopyrightInformation.html

Concept ACK

**Comment by laanwj** [comment]
>  My initial resistance was due to me misunderstanding that I was being asked to assign my copyright to somebody else.

You cannot 'accidentally' assign your copyright to someone else, only by signing an explicit contract. Some projects (such as leveldb) have a separate contributor license agreement that transfers copyright to a company or organization (to Google in that case). Bitcoin Core does not have that.
The only thing you implicitly agree on by contributing is licensing the code under the license of the project (so, MIT).

**Comment by MarcoFalke** [comment]
Anyone knows what is the proper way to test this on fedora with just
self-compiled Bitcoin Core?

**Comment by laanwj** [comment]
> Anyone knows what is the proper way to test this on fedora with just self-compiled Bitcoin Core?

My guess would be typing this into your shell:
```bash
export PATH=$PATH:$BUILDDIR/src
source contrib/bitcoin-tx.bash-completion
source contrib/bitcoind.bash-completion
```
The completion script are simply bash scripts that define a function and install it using `complete`.
But it doesn't seem to work, it complains about not being able to find the command `have`.

**Comment by roques** [comment]
I've looked into testing the completions without installing them and found that the use of `have` is deprecated. The now updated bash-completion snippets don't do that anymore and when using `dh_bash-completion` from bash-completion ≥ 1:2.1-4.2 while building the Debian package will be installed in `/usr/share/bash-completion/completions/` instead of the deprecated `/etc/bash_completion.d/` and autoloaded ondemand.
@MarcoFalke, for you this should mean that when you have Fedora's `bash-completion` installed and active, you should now be able to simply source the scripts to test them without installing them.

**Comment by MarcoFalke** [comment]
Appears to be working:


```
$ src/qt/bitcoin-qt -
Display all 101 possibilities? (y or n)
-?                    -datacarrier          -loadblock=           -onlynet=             -rpcallowip=          -torcontrol=
-addnode=             -datacarriersize      -logips               -par=                 -rpcauth=             -torpassword=
-alertnotify=         -datadir=             -logtimestamps        -paytxfee=            -rpcbind=             -txconfirmtarget=
-banscore=            -dbcache=             -maxconnections=      -peerbloomfilters     -rpccookiefile=       -txindex
-bantime=             -debug=               -maxmempool=          -permitbaremultisig   -rpcpassword=         -uacomment=
-bind=                -disablewallet        -maxorphantx=         -pid=                 -rpcport=             -upgradewallet
-blockmaxcost=        -discover             -maxreceivebuffer=    -port=                -rpcthreads=          -usehd
-blockmaxsize=        -dns                  -maxsendbuffer=       -printtoconsole       -rpcuser=             -version
-blockminsize=        -dnsseed              -maxtimeadjustment    -proxy=               -salvagewallet        -wallet=
-blocknotify=         -externalip=          -maxtxfee=            -proxyrandomize       -seednode=            -walletbroadcast
-blockprioritysize=   -fallbackfee=         -maxuploadtarget=     -prune=               -server               -walletnotify=
-bytespersigop        -forcednsseed         -mempoolexpiry=       -reindex              -shrinkdebugfile      -whitebind=
-checkblocks=         -help-debug           -mempoolreplacement   -reindex-chainstate   -spendzeroconfchange  -whitelist=
-checklevel=          -keypool=             -min                  -rescan               -splash               -whitelistforcerelay
-choosedatadir        -lang=                -minrelaytxfee=       -resetguisettings     -sysperms             -whitelistrelay
-conf=                -listen               -mintxfee=            -rest                 -testnet              -zapwallettxes=
-connect=             -listenonion          -onion=               -rootcertificates=    -timeout=             
```


Though, I couldn't get it work with the cli:

```
$ source contrib/bitcoin-cli.bash-completion
$ src/bitcoin-cli -regtest g[tab][tab]
```
shows nothing

**Comment by laanwj** [comment]
@MarcoFalke Seems to work here, do you have the daemon running? I think it queries the remote for the command list:
```
(8289)$ src/bitcoin-cli -regtest ge↹
generate               getaddressesbyaccount  getblockcount          getconnectioncount     getmempoolentry        getnetworkinfo         getrawtransaction      gettxoutproof
generatetoaddress      getbalance             getblockhash           getdifficulty          getmempoolinfo         getnewaddress          getreceivedbyaccount   gettxoutsetinfo
getaccount             getbestblockhash       getblockheader         getinfo                getmininginfo          getpeerinfo            getreceivedbyaddress   getunconfirmedbalance
getaccountaddress      getblock               getblocktemplate       getmempoolancestors    getnettotals           getrawchangeaddress    gettransaction         getwalletinfo
getaddednodeinfo       getblockchaininfo      getchaintips           getmempooldescendants  getnetworkhashps       getrawmempool          gettxout          
```

**Comment by laanwj** [comment]
ACK 1ba3db6

**Comment by MarcoFalke** [comment]
I am not running the daemon. This is good for merge.

**Pull request merged**